source/game/kart/KartMove.cc

```diff
@@ -1593,6 +1624,146 @@ void KartMove::landTrick() {
     activateBoost(KartBoost::Type::TrickAndZipper, duration);
 }
 
+/// @addr{0x8058498C}
+void KartMove::enterCannon() {
+    init(true, true);
+    physics()->clearDecayingRot();
+    m_boost.resetActive();
+    state()->setBoost(false);
+
+    cancelJumpPad();
+    m_rampBoost = 0;
+    state()->setRampBoost(false);
+    m_ssmtCharge = 0;
+    m_bPadBoost = false;
+    m_bRampBoost = false;
+    m_bPadJump = false;
+    m_bSsmtCharged = false;
+    m_bSsmtLeeway = false;
+    m_bTrickableSurface = false;
+    m_bWallBounce = false;
+    m_offroadInvincibility = 0;
+    state()->setBoostOffroadInvincibility(false);
+
+    dynamics()->reset();
+
+    state()->setHop(false);
+    state()->setInCannon(true);
+    state()->setSkipWheelCalc(true);
+    state()->setCannonStart(false);
+
+    const auto [cannonPos, cannonDir] = cannonPosRot();
+    m_cannonEntryPos = pos();
+    m_cannonEntryOfs = cannonPos - pos();
+    m_cannonEntryOfsLength = m_cannonEntryOfs.normalise();
+    m_cannonEntryOfs.normalise();
+    m_dir = m_cannonEntryOfs;
+    m_vel1Dir = m_cannonEntryOfs;
+    m_cannonOrthog = EGG::Vector3f::ey.perpInPlane(m_cannonEntryOfs, true);
+    m_cannonProgress.setZero();
+}
+
+/// @addr{0x80584D58}
+void KartMove::calcCannon() {
+    auto [cannonPos, cannonDir] = cannonPosRot();
+    EGG::Vector3f forwardXZ = cannonPos - m_cannonEntryPos - m_cannonProgress;
+    EGG::Vector3f forward = forwardXZ;
+    f32 forwardLength = forward.normalise();
+    forwardXZ.y = 0;
+    forwardXZ.normalise();
+    EGG::Vector3f local94 = m_cannonEntryOfs;
+    local94.y = 0;
+    local94.normalise();
+    m_speedRatioCapped = 1.0f;
+    EGG::Matrix34f cannonOrientation;
+    cannonOrientation.makeOrthonormalBasis(forward, EGG::Vector3f::ey);
+    EGG::Vector3f up = cannonOrientation.multVector33(EGG::Vector3f::ey);
+    m_smoothedUp = up;
+    m_up = up;
+
+    // exiting cannon
+    if ((forwardLength < 30.0f || local94.dot(forwardXZ) <= 0.0f) && state()->isInCannon()) {
+        state()->setInCannon(false);
+        state()->setSkipWheelCalc(false);
+        dynamics()->setIntVel(m_speed * m_cannonEntryOfs);
+        return;
+    }
+
+    cannonOrientation.mulRow33(0, EGG::Vector3f::ez);
+    cannonOrientation.mulRow33(1, EGG::Vector3f::ez);
+    m_speed = m_baseSpeed;
+    const auto *cannonPoint =
+            System::CourseMap::Instance()->getCannonPoint(state()->cannonPointId());
+    s16 cannonParameterIdx = cannonPoint->parameterIdx();
+    if (cannonParameterIdx < 1) {
+        cannonParameterIdx = 0;
+    }
+    assert(static_cast<size_t>(cannonParameterIdx) < CANNON_PARAMETERS.size());
+    const auto cannonParams = CANNON_PARAMETERS[cannonParameterIdx];
```

Comment:
```suggestion
    const auto &cannonParams = CANNON_PARAMETERS[cannonParameterIdx];
```
This creates a copy otherwise.